examples/python/Boxes_QPE_example.py

```diff
@@ -0,0 +1,543 @@
+
+
+# ## State Preparation
+
+# Lets now consider a more interesting Hamiltonian. We will look at the $H_2$ Hamiltonian for a bond length of 5 angstroms.
+#
+# We can define the Hamiltonian as a `pytket` `QubitPauliOperator` and then synthesise a circuit for the time evolved Hamiltonian with the `gen_term_sequence_circuit` utility method.
+#
+# Here we can load in our `QubitPauliOperator` from a JSON file.
+
+# In[22]:
+
+
+import json
+from pytket.utils import QubitPauliOperator
+
+with open("h2_5A.json") as f:
+    qpo_h25A = QubitPauliOperator.from_list(json.load(f))
+
+
+# In[23]:
+
+
+ham_circ = gen_term_sequence_circuit(qpo_h25A, Circuit(4))
+
+
+# In[24]:
+
+
+from pytket.passes import DecomposeBoxes
+
+
+# In[25]:
+
+
+DecomposeBoxes().apply(ham_circ)
+
+
+# In[26]:
+
+
+render_circuit_jupyter(ham_circ)
+
+
+# Now have to come up with an ansatz state to feed into our phase estimation algorithm.
+#
+# We will use the following ansatz
+#
+# $$
+# \begin{equation}
+# |\psi_0\rangle = e^{i \frac{\pi}{2}\theta YXXX}|1100\rangle\,.
+# \end{equation}
+# $$
+#
+# We can synthesise a circuit for the Pauli exponential using `PauliExpBox`.
+
+# In[27]:
+
+
+from pytket.pauli import Pauli
+from pytket.circuit import PauliExpBox
+
+state_circ = Circuit(4).X(0).X(1)
+yxxx = [Pauli.Y, Pauli.X, Pauli.X, Pauli.X]
+yxxx_box = PauliExpBox(yxxx, 0.1)  # here theta=0.1
+state_circ.add_gate(yxxx_box, [0, 1, 2, 3])
+
+
+# we can extract the statevector for $e^{i \frac{\pi}{2}\theta YXXX}|1100\rangle$ using the `Circuit.get_statvector()` method.
+
+# In[28]:
+
+
+initial_state = state_circ.get_statevector()
+
+
+# Finally we can prepare our initial state using `StatePreparationBox`.
+
+# In[29]:
+
+
+from pytket.circuit import StatePreparationBox
+
+state_prep_box = StatePreparationBox(initial_state)
+
+
+# We now have all of the ingredients we need to build our phase estimation circuit for the $H_2$ molecule. Here we will use 8 qubits to estimate the phase.
+#
+# Note that the number of controlled unitaries in our circuit will scale exponentially with the number of measurement qubits. Decomposing these controlled boxes to native gates can lead to very deep circuits.
+#
+# We will again use the idealised `AerBackend` simulator for our simulation. If we were instead using a simulator with a noise model or a NISQ device we would expect our results to be degraded due to the large circuit depth.
+
+# In[30]:
+
+
+ham_state_prep_circuit = Circuit(4).add_gate(state_prep_box, [0, 1, 2, 3])
+
+
+# In[31]:
+
+
+h2_qpe_circuit = build_phase_est_circuit(
+    8, state_prep_circuit=ham_state_prep_circuit, unitary_circuit=ham_circ
+)
+
+
+# In[32]:
+
+
+render_circuit_jupyter(h2_qpe_circuit)
+
+
+# In[33]:
+
+
+compiled_ham_circ = backend.get_compiled_circuit(h2_qpe_circuit, 0)
+
+
+# In[34]:
+
+
+n_shots = 2000
+
+ham_result = backend.run_circuit(compiled_ham_circ, n_shots=n_shots)
+
+
+# In[35]:
+
+
+plot_qpe_results(ham_result, y_limit=1.2 * n_shots, n_strings=5)
```

Comment:
I should probably give a bit of an explanation of the results here but I don't see a straightforward relation between the eigenvalues of H and my measurement results.